packages/medusa/src/services/cart.ts

```diff
@@ -1631,43 +1632,46 @@ class CartService extends TransactionBaseService {
   }
 
   /**
-   * Sets a payment method for a cart.
+   * Selects a payment session for a cart and creates a payment object in the external provider system
    * @param cartId - the id of the cart to add payment method to
    * @param providerId - the id of the provider to be set to the cart
-   * @return result of update operation
    */
-  async setPaymentSession(cartId: string, providerId: string): Promise<Cart> {
+  async setPaymentSession(cartId: string, providerId: string): Promise<void> {
     return await this.atomicPhase_(
       async (transactionManager: EntityManager) => {
         const psRepo = transactionManager.getCustomRepository(
           this.paymentSessionRepository_
         )
 
         const cart = await this.retrieveWithTotals(cartId, {
-          relations: ["region", "region.payment_providers", "payment_sessions"],
+          relations: [
+            "customer",
+            "region",
+            "region.payment_providers",
+            "payment_sessions",
+          ],
         })
 
-        // The region must have the provider id in its providers array
-        if (
-          providerId !== "system" &&
-          !(
-            cart.region.payment_providers.length &&
-            cart.region.payment_providers.find(({ id }) => providerId === id)
-          )
-        ) {
+        const isProviderPresent = cart.region.payment_providers.find(
+          ({ id }) => providerId === id
+        )
+
+        if (providerId !== "system" && !isProviderPresent) {
           throw new MedusaError(
             MedusaError.Types.NOT_ALLOWED,
             `The payment method is not available in this region`
           )
         }
 
-        await Promise.all(
-          cart.payment_sessions.map(async (paymentSession) => {
-            return psRepo.save({ ...paymentSession, is_selected: null })
-          })
+        const cartPaymentSessionIds = cart.payment_sessions.map((p) => p.id)
+        await psRepo.update(
+          { id: In(cartPaymentSessionIds) },
+          {
+            is_selected: null,
```

Comment:
question: In the data model, it says that the `is_selected` column is `NOT NULL`. Why is this passing in null here?

Reply:
<img width="640" alt="Screenshot 2023-01-09 at 11 56 09" src="https://user-images.githubusercontent.com/25098370/211292498-6e0443c7-a9ba-4c26-bb83-e8cb161eb099.png">
I think there is confusion between is_selected and is_initiated right?

Reply:
is_selected is also nullable in the db

Reply:
ah, fair. You caught it correctly. Got confused between the 2 😓